src/smallmap_gui.cpp

```diff
@@ -594,1113 +611,1229 @@ static const byte _vehicle_type_colours[6] = {
 
-	this->SetWidgetLoweredState(WID_SM_TOGGLETOWNNAME, this->show_towns);
+		this->RebuildColourIndexIfNecessary();
 
-	this->SetupWidgetData();
+		this->SetWidgetLoweredState(WID_SM_SHOW_HEIGHT, _smallmap_show_heightmap);
 
-	this->SetZoomLevel(ZLC_INITIALIZE, nullptr);
-	this->SmallMapCenterOnCurrentPos();
-	this->SetOverlayCargoMask();
-}
+		this->SetWidgetLoweredState(WID_SM_TOGGLETOWNNAME, this->show_towns);
 
-SmallMapWindow::~SmallMapWindow()
-{
-	delete this->overlay;
-}
+		this->SetupWidgetData();
 
-/* virtual */ void SmallMapWindow::Close([[maybe_unused]] int data)
-{
-	this->BreakIndustryChainLink();
-	this->Window::Close();
-}
+		this->SetZoomLevel(ZLC_INITIALIZE, nullptr);
+		this->SmallMapCenterOnCurrentPos();
+		this->SetOverlayCargoMask();
+	}
 
-/**
- * Rebuilds the colour indices used for fast access to the smallmap contour colours based on the heightlevel.
- */
-void SmallMapWindow::RebuildColourIndexIfNecessary()
-{
-	/* Rebuild colour indices if necessary. */
-	if (SmallMapWindow::map_height_limit == _settings_game.construction.map_height_limit) return;
+	virtual ~SmallMapWindow()
+	{
+		delete this->overlay;
+	}
 
-	for (uint n = 0; n < lengthof(_heightmap_schemes); n++) {
-		/* The heights go from 0 up to and including maximum. */
-		int heights = _settings_game.construction.map_height_limit + 1;
-		_heightmap_schemes[n].height_colours = ReallocT<uint32_t>(_heightmap_schemes[n].height_colours, heights);
+	/**
+	 * Center the small map on the current center of the viewport.
+	 */
+	void SmallMapCenterOnCurrentPos()
+	{
+		const Viewport *vp = GetMainWindow()->viewport;
+		Point viewport_center = InverseRemapCoords2(vp->virtual_left + vp->virtual_width / 2, vp->virtual_top + vp->virtual_height / 2);
 
-		for (int z = 0; z < heights; z++) {
-			size_t access_index = (_heightmap_schemes[n].colour_count * z) / heights;
+		int sub;
+		const NWidgetBase *wid = this->GetWidget<NWidgetBase>(WID_SM_MAP);
+		Point sxy = this->ComputeScroll(viewport_center.x / (int)TILE_SIZE, viewport_center.y / (int)TILE_SIZE,
+				std::max(0, (int)wid->current_x / 2 - 2), wid->current_y / 2, &sub);
+		this->SetNewScroll(sxy.x, sxy.y, sub);
+		this->SetDirty();
+	}
 
-			/* Choose colour by mapping the range (0..max heightlevel) on the complete colour table. */
-			_heightmap_schemes[n].height_colours[z] = _heightmap_schemes[n].height_colours_base[access_index];
-		}
+	/**
+	 * Get the center of the given station as point on the screen in the smallmap window.
+	 * @param st Station to find in the smallmap.
+	 * @return Point with coordinates of the station.
+	 */
+	Point GetStationMiddle(const Station *st) const
+	{
+		int x = CenterBounds(st->rect.left, st->rect.right, 0);
+		int y = CenterBounds(st->rect.top, st->rect.bottom, 0);
+		Point ret = this->RemapTile(x, y);
+
+		/* Same magic 3 as in DrawVehicles; that's where I got it from.
+		* No idea what it is, but without it the result looks bad.
```

Comment:
Indentation is broken for the comments. Here and in many other places as well.
`grep -P "[\t]\*[ /]" smallmap_gui.cpp`